tool/tctl/common/bots_command.go

```diff
@@ -546,6 +526,296 @@ func (c *BotsCommand) UpdateBot(ctx context.Context, client *authclient.Client)
+	var instances []*machineidv1pb.BotInstance
+	req := &machineidv1pb.ListBotInstancesRequest{}
+
+	if c.botName != "" {
+		req.FilterBotName = c.botName
+	}
+
+	for {
+		resp, err := client.BotInstanceServiceClient().ListBotInstances(ctx, req)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		instances = append(instances, resp.BotInstances...)
+		if resp.NextPageToken == "" {
+			break
+		}
+		req.PageToken = resp.NextPageToken
+	}
+
+	if c.format == teleport.JSON {
+		err := utils.WriteJSONArray(c.stdout, instances)
+		if err != nil {
+			return trace.Wrap(err, "failed to marshal bot instances")
+		}
+
+		return nil
+	}
+
+	if len(instances) == 0 {
+		if c.botName == "" {
+			fmt.Fprintln(c.stdout, "No bot instances found.")
+		} else {
+			fmt.Fprintf(c.stdout, "No bot instances found with name %q.\n", c.botName)
+		}
+		return nil
+	}
+
+	t := asciitable.MakeTable([]string{"ID", "Join Method", "Hostname", "Joined", "Last Seen", "Generation"})
+	for _, i := range instances {
+		var (
+			joinMethod string
+			hostname   string
+			generation string
+		)
+
+		joined := i.Status.InitialAuthentication.AuthenticatedAt.AsTime().Format(time.RFC3339)
+		initialJoinMethod := i.Status.InitialAuthentication.JoinMethod
+
+		lastSeen := i.Status.InitialAuthentication.AuthenticatedAt.AsTime()
+
+		if len(i.Status.LatestAuthentications) == 0 {
+			generation = "n/a"
+		} else {
+			auth := i.Status.LatestAuthentications[len(i.Status.LatestAuthentications)-1]
+
+			generation = fmt.Sprint(auth.Generation)
+
+			if auth.JoinMethod == initialJoinMethod {
+				joinMethod = auth.JoinMethod
+			} else {
+				// If the join method changed, show the original method and latest
+				joinMethod = fmt.Sprintf("%s (%s)", auth.JoinMethod, initialJoinMethod)
+			}
+
+			if auth.AuthenticatedAt.AsTime().After(lastSeen) {
+				lastSeen = auth.AuthenticatedAt.AsTime()
+			}
+		}
+
+		if len(i.Status.LatestHeartbeats) == 0 {
+			hostname = "n/a"
+		} else {
+			hb := i.Status.LatestHeartbeats[len(i.Status.LatestHeartbeats)-1]
+
+			hostname = hb.Hostname
+
+			if hb.RecordedAt.AsTime().After(lastSeen) {
+				lastSeen = hb.RecordedAt.AsTime()
+			}
+		}
+
+		t.AddRow([]string{
+			fmt.Sprintf("%s/%s", i.Spec.BotName, i.Spec.InstanceId), joinMethod,
+			hostname, joined, lastSeen.Format(time.RFC3339), generation,
+		})
+	}
+	fmt.Fprintln(c.stdout, t.AsBuffer().String())
+
+	fmt.Fprintf(c.stdout, "\nTo view more information on a particular instance, run:\n\n> %s bots instances show [id]\n", os.Args[0])
+
+	if c.botName != "" {
+		fmt.Fprintf(c.stdout, "\nTo onboard a new instance for this bot, run:\n\n> %s bots instances add %s\n", os.Args[0], c.botName)
+	}
+
+	return nil
+}
+
+// AddBotInstance begins onboarding a new instance of an existing bot.
+func (c *BotsCommand) AddBotInstance(ctx context.Context, client *authclient.Client) error {
+	// A bit of a misnomer but makes the terminology a bit more consistent. This
```

Comment:
I tried to think of something better than "tctl bots instances add" but I think it's close enough and not too confusing. I figure the fact that it basically just outputs an example command for delegated join methods makes sense in those scenarios.

Reply:
Yep, I don't love it but I think it's memorable and at least conceptually similar to existing commands in tctl.